src/feature/hs/hs_dos.c

```diff
@@ -7,14 +7,84 @@
 
 #define HS_DOS_PRIVATE
 
+#include "core/or/or.h"
+#include "app/config/config.h"
+
 #include "core/or/circuitlist.h"
 
+#include "feature/nodelist/networkstatus.h"
+#include "feature/relay/routermode.h"
+
+#include "lib/evloop/token_bucket.h"
+
 #include "hs_dos.h"
 
+/* Default values. */
+#define HS_DOS_INTRODUCE_CELL_RATE_PER_SEC 25
+#define HS_DOS_INTRODUCE_CELL_BURST_PER_SEC 200
+
+/* Consensus parameters. */
+static uint32_t hs_dos_introduce_rate_per_sec;
+static uint32_t hs_dos_introduce_burst_per_sec;
+
+/* Return the parameter for the introduction rate per sec. */
+static uint32_t
+get_param_rate_per_sec(const networkstatus_t *ns)
+{
+  return networkstatus_get_param(ns, "HSDoSIntroduceRate",
+                                 HS_DOS_INTRODUCE_CELL_RATE_PER_SEC,
+                                 0, INT32_MAX);
+}
+
+/* Return the parameter for the introduction burst per sec. */
+static uint32_t
+get_param_burst_per_sec(const networkstatus_t *ns)
+{
+  return networkstatus_get_param(ns, "HSDoSIntroduceBurst",
+                                 HS_DOS_INTRODUCE_CELL_BURST_PER_SEC,
+                                 0, INT32_MAX);
+}
+
+/* Set consensus parameters. */
+static void
+set_consensus_parameters(const networkstatus_t *ns)
+{
+  hs_dos_introduce_rate_per_sec = get_param_rate_per_sec(ns);
+  hs_dos_introduce_burst_per_sec = get_param_burst_per_sec(ns);
+}
+
 /*
  * Public API.
  */
 
+/* Return the INTRODUCE2 cell rate per second. */
```

Comment:
I'm anxious that these getters will get called before `set_consensus_parameters()` gets called and they will return 0, resulting in an all-rejecting token bucket. Can we get some assurance than when the getters get called, they have been intiialized or they return the default val? What happens if we merge `hs_dos_get_intro2_rate()` and `get_param_rate_per_sec()` are you worried we are putting too much CPU on the fast path?

Reply:
Here's one quick way to ensure that we never read zero-initialised values: initialise the static variables with the default values.

Reply:
Fixup commit: 118cbd4319

Reply:
Ah that looks good. Perhaps we can also change the name of these constants to `HS_DOS_INTRODUCE_DEFAULT_CELL_RATE_PER_SEC` to denote that they are default values and not the final truth?

Reply:
Commit 49c9f21255]